lib/elixir/lib/kernel/error_handler.ex

```diff
@@ -3,28 +3,36 @@
 defmodule Kernel.ErrorHandler do
   @moduledoc false
 
-  def undefined_function(module, fun, args) do
+  @spec undefined_function(module, atom, list) :: term
+  def undefined_function(module, fun, args)
+      when is_atom(module) and is_atom(fun) and is_list(args) do
     ensure_loaded(module) or ensure_compiled(module, :module)
     :error_handler.undefined_function(module, fun, args)
   end
 
-  def undefined_lambda(module, fun, args) do
+  @spec undefined_function(module, fun, list) :: term
+  def undefined_lambda(module, fun, args)
+      when is_atom(module) and is_function(fun) and is_list(args) do
     ensure_loaded(module) or ensure_compiled(module, :module)
     :error_handler.undefined_lambda(module, fun, args)
   end
 
-  def ensure_loaded(module) do
+  @spec ensure_loaded(module) :: boolean
+  def ensure_loaded(module) when is_atom(module) do
     case :code.ensure_loaded(module) do
       {:module, _} -> true
       {:error, _} -> false
     end
   end
 
+  @spec ensure_compiled(module, atom) :: boolean
+  @spec ensure_compiled(nil, any) :: false
```

Comment:
I wouldn't add this clause because it doesn't add information for consumers of this API, at the documentation level (even though it is technically more correct).